### PR TITLE
CAD-3444 handshake changes from p2p-master

### PR DESCRIPTION
**Pull request by coot**
- handshake codec
- handshake: simplify types of client / server
- handshake: resiliant for simultanous open
- handshake: refactor the codec
- handshake: added test for a simultaneous open connections
- handshake: updated network-spec
- handshake: remove versioned application from HandshakeArguments

### CODE REVIEW

**Review by karknu — APPROVED** [review]
LGTM.

**Comment by coot** [comment]
bors merge

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs [hunk 1]
```diff
@@ -82,6 +82,16 @@ instance Protocol (Handshake vNumber vParams) where
         :: Map vNumber vParams
         -> Message (Handshake vNumber vParams) StPropose StConfirm
 
+      -- |
+      -- `MsgProposeVersions'` received as a response to 'MsgProposeVersions'.
+      -- It is not supported to explicitly send this message. It can only be
+      -- received as a copy of 'MsgProposeVersions' in a simultanous open
+      -- scenario.
+      --
+      MsgProposeVersions'
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs:91]: `MsgProposeVersionsSimultaneousOpen`

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs [hunk 2]
```diff
@@ -111,11 +121,21 @@ instance Protocol (Handshake vNumber vParams) where
...
 
+-- $simultanous-open
+--
+-- On simultanous open both sides will send `MsgProposeVersions`, which will be
+-- decoded as `MsgProposeVersions'` which is a terminal message of the
+-- protocol.  It is important to stress that in this case both sides will make
+-- the choice which version and parameters to pick.  Our algorithm for picking
+-- version is symmetric, which ensures that both sides will endup with the same
+-- choice.  If one side decides to refuse the version it will close the
+-- connection, without sending the reason to the other side.
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs:132]: TODO: 

* add a property test for the function that picks the version,  it must be symmetric,
* document this for 3rd parties.

Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs:132]: Elaborate on the consequences of non injective encoding and reason why this is ok here.

File: docs/network-spec/miniprotocols.tex [hunk 3]
```diff
@@ -359,6 +359,7 @@ \section{Handshake mini-protocol}
 \newcommand{\StPropose}{\state{StPropose}}
 \newcommand{\StConfirm}{\state{StConfirm}}
 \newcommand{\MsgProposeVersions}{\msg{MsgProposeVersions}}
+\newcommand{\MsgProposeVersionsX}{\msg{MsgProposeVersions'}}
```
Comment (coot) [inline docs/network-spec/miniprotocols.tex:362]: Change the name of the `MsgProposeVersions'`.

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Type.hs:132]: done

File: docs/network-spec/miniprotocols.tex (hunk 3 above)
Comment (coot) [reply docs/network-spec/miniprotocols.tex:362]: done

**Pull request merged**